### ISSUE TITLE
Angular Adapter

### ISSUE THREAD

**Issue by pit999** [ivanhofer/typesafe-i18n#62]
A adapter for Angular or setup/sample for Angular would be helpful.

**Comment by ivanhofer** [comment]
Hi @pit999, I'm no Angular expert, so it's hard for me to provide a good solution. Would you like to help contribute?

**Comment by pit999** [comment]
Yes :-). 
I tried the node adapter but got some error messages when I compile/run the angular app.
After that I tried with the undefined adapter and now I have to understand what to do.
It was so easy to use it with node without understanding the code in detail :-)

**Comment by ivanhofer** [comment]
All you need should be inside the generated `src/i18n/typesafe-util.ts` file. Basically you would need to write a wrapper (maybe an Angular Service) for these functions, that handles your state (current selected locale, translation object, etc.).

If you take a look at the existing adapters for [react](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/adapter-react/src/react-adapter.tsx) or [svelte](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/adapter-svelte/src/svelte-adapter.ts) you will see that they provide some useful functions, initialize and store the translationObject `i18nObject`.

In these files you will see a lot of TypeScript generics. You could replace them with `any` while testing. If you can make it wok for your Angular Application I can add the correct types later on. 
> You should set the option "loadLocalesAsync" to `false` in order to make the whole process less complex. We can make it work for async later on

**Comment by pit999** [comment]
Thank you. Will try that and keep you informed.

**Comment by pit999** [comment]
It will take some time. I never developed an Angular Module so far (so I have to learn that too) and to dig deeper into your code.
At least I found an Angular Module implementation for i18next (https://github.com/Romanchuk/angular-i18next-demo). I could use this module, but I like the type safe implementation.

**Comment by pit999** [comment]
Above link was for a demo. The module is here: https://github.com/Romanchuk/angular-i18next

**Comment by ivanhofer** [comment]
I think a Service would be the preferred solution. As far as I can remember, Modules are more complex to integrate into another application than using only a service. I haven't written any angular code in years, so maybe that has changed. Also from the complexity point, I think the whole angular integration should  only take < 100 lines of code.

**Comment by pit999** [comment]
I see some benefits for a module (easy to reuse/install, use of pipes which makes the overall handling much easier (otherwise you a variable in your component class to use in the html file) and the overhead ist not that much - as far as I can see :-)
And I'm not under time pressure :-)
Did you use Angular 1 or >2?

**Comment by ivanhofer** [comment]
Sounds you know a lot more of Angular like I do ;)
I think it was version 4. But I heard that in recent years they improved the developer experience a lot, so I assume some parts have changed a lot.
I'm looking forward for the solution you will come up with :)

**Comment by pit999** [comment]
Additionally I tried just to use detectLocale in an Angular class:
import { localStorageDetector } from 'typesafe-i18n/detectors'
import { detectLocale } from './../i18n/i18n-util'
...
const detectedLocale = detectLocale(localStorageDetector);

But get two errors:
ERROR in ./src/i18n/i18n-util.ts 21:46-60
"export 'detectLocale' (imported as 'detectLocaleFn') was not found in 'typesafe-i18n/detectors'

ERROR in ./src/app/app.component.ts 18:44-64
"export 'localStorageDetector' was not found in 'typesafe-i18n/detectors'

**Comment by ivanhofer** [comment]
Hi @pit999,
the parts inside the `< >` are called generics. You can read about them here: https://www.typescriptlang.org/docs/handbook/2/generics.html
Basically it is a way to model a function that can take different arguments and the output of that function depends on the input types. In this case e.g. if we pass in `'en' | 'de'` as the first generic (`L`), typescript will yell at you if you try to pass in `'ru'` for the `locale` argument in the `setLocale` function.
You should read about it in the link above. The concept is explained with some basic examples.

The types in the adapters can be complicated for beginners, but the generics are the reason why `typesafe-i18n` can offer its typesafety features. You could also ignore all types for now and pass `any` when typescript yells at you. I can add the types later on. The first step would be to come up with a concept that works within Angular. 

About the error: you should install the latest version. The issue should be fixed there.

**Comment by pit999** [comment]
Thank you, will read the article.

version 2.33 fixed the problem. 

Now I tried to use the node implementation and get following error during run time.
i18n-util.ts:34 Uncaught TypeError: Object(...) is not a function
    at i18n (i18n-util.ts:34)
    at Module.9amk (i18n-node.ts:6)
export const i18n = () => initI18n<Locales, Translation, TranslationFunctions, Formatters>(getTranslationForLocale, initFormatters)

Here I need the wrapper :-)

**Comment by pit999** [comment]
I read the article yesterday. For normal generics I understand the concept, but this one looks different :-) I will try with any. 
One last question for now: Should I work with L, LL or LLL? (A newbie to i18n too :-))

**Comment by ivanhofer** [comment]
@pit999 I'm not sure why you get this error. Please only use the function inside `i18n-utils.ts`. These should work.

These functions (`L`, `LL` and `LLL`) are not common for i18n. In `typesasfe-i18n` translation functions are named like this.
You can read more about the differences [here](https://github.com/ivanhofer/typesafe-i18n#general).

You should probably work with `LL`.
`L` is a wrapper around `LL` which is a wrapper around `LLL`.
`L` is mostly for APIs where you need to access all languages the same time.
`LL` is for frontend-applications where only a single language will be shown to a user.
`LLL` is for niche cases where you only want to translate single strings.

**Comment by ivanhofer** [comment]
@pit999 I have created a new branch [`angular`](https://github.com/ivanhofer/typesafe-i18n/tree/angular) and placed a basic Angular app into to [`examples`](https://github.com/ivanhofer/typesafe-i18n/tree/angular/examples/angular) folder.
Maybe you can come up wit a better solution on how `typesafe-i18n` should work inside a Angular application?

**Comment by pit999** [comment]
Really cool and amazing :-) I installed your code and run it successfully.

I adapted my code and still get the following error in
i18nService.initI18n('en').then(() => {
async initI18n(locale: Locales): Promise<void> {
		await this.setLocale(locale)
	}
async setLocale(locale: Locales): Promise<void> {
		**this._LL = await i18nObject(locale)**
		this.currentLocale = locale
	}

core.js:4442 ERROR Error: Uncaught (in promise): TypeError: Object(...) is not a function
TypeError: Object(...) is not a function
    at i18nObject (i18n-util.ts:28)
    at I18nService.<anonymous> (i18n.service.ts:30)

I compared every file in your and my code - everything looks equal.
Updated to "typesafe-i18n": "^2.34.0" too.
Wired, will look into it later...

**Comment by pit999** [comment]
And thanks for your support, amazing. I'm on holiday next week, but will look into the module implementation as I see great benefit in using pipes in the html files :-)

**Comment by ivanhofer** [comment]
@pit999 what Angular version do you use?
Angular sometimes wraps native browser functions. I had to implement a workaround in order to make the asynchronous code work with Angular because of that.
So maybe there is a bug somewhere in the Angular implementation?

Could you please share the code-snippet that is on line 4442 of `core.js`?

**Comment by pit999** [comment]
I'm using Angular 10.
core.js in angular\core?
I found these lines in fesm2015 \core.js (4441-4443
function defaultErrorLogger(console, ...values) {
    console.error(...values);
}

**Comment by pit999** [comment]
Line 4442 of angular12\core.js is in angular10\core.js
const ES2015_INHERITED_CLASS_WITH_DELEGATE_CTOR = /^class\s+[A-Za-z\d$_]*\s*extends\s+[^{]+{[\s\S]*constructor\s*\(\)\s*{\s*super\(\.\.\.arguments\)/;

**Comment by ivanhofer** [comment]
hmm.. i can't really see what causes the error ` TypeError: Object(...) is not a function`. Maybe you could set a breakpoint and see what happens?

**Comment by pit999** [comment]
In my app:
![image](https://user-images.githubusercontent.com/5500277/125920738-bb42848a-0b4b-4f93-a66d-5e25bdabc94f.png)

In your app:
![image](https://user-images.githubusercontent.com/5500277/125920864-02e965c6-9f70-4698-814c-39ea8254876c.png)

**Comment by ivanhofer** [comment]
@pit999 I have found the issue and released version `2.35.1`. It should now also work with older Angular versions.

**Comment by pit999** [comment]
Great, I tried and it works now in my app. Will check in more detail after my holidays :-)
Thanks for your help so far. I'm sure I never got this running :-)

**Comment by pit999** [comment]
Hi Ivan, sorry took a while to look deeper into this to develop an Angular pipe.
I got a simple pipe now running, but face a problem.
An Angular pipe provides a key and options you can use in the pipe.
So it's not static. Is there an option to use the key to access the related transform method of i18n typesafe?
I know it's a bit ridiculous, because you can use any key in the html tag with the pipe and no more typesafe.
But the alternative would be to define for any string a variable in the component and use use this variable in the html file.
Any other ideas?

  public transform(**key: string** | string[], options?: PipeOptions): string {
    options = this.prepareOptions(options);

    let result: string = this.i18nService.**LL.delete()**;

    return result;
  }

**Comment by ivanhofer** [comment]
Hi @pit999,

I'm not sure if a Pipe would be the best option here..
How would you call this in your template code?

When we would have a translation
```typescript
const en: BaseTranslation = {
   HI: 'Hello {name:string}'
}
```
I only came up with something like this:
```
{ 'HI' | I18n:{ name: 'John' } }
```
This looks odd and I don't see any advantages over using it like it currently is
```
{{ LL.HI({ name: 'John' }) }}
```

Maybe you can tell how you would like to use it in a Angular application?

**Comment by pit999** [comment]
Hi,
Your solution is absolutely fine with me :-)
I was not aware that this is possible.
I will use typesafe now in all my components to support different languages.
Best regards and thanks a lot,
Peter

**Comment by pit999** [comment]
Hi,

another question: The following is not possible, isn't it?
{{ LL.HI({ entity: {{LL.entityName()}} }) }} or {{ LL.HI({ entity: LL.entityName() }) }}

**Comment by pit999** [comment]
I think I do not need the component in my components.
I changed the code to:
import { I18nService } from 'src/i18n/i18n.service';
...
  i18LL: any;

  constructor(
    i18nService: I18nService
  ) {
    this.i18LL = i18nService.LL;

and use e.g. {{i18LL.status()}} in the html file.

**Comment by ivanhofer** [comment]
Ok great!
Please test it in your application and let me know if everything works. Then I can add an 'angular' adapter option to the generator.

Using only the service would mean that every component that uses i18n would need to have following code implemented
```typescript
@Component({ /* ... */ })
export class SomeComponent {

  constructor(private i18nService: I18nService) {  }

  get LL(): TranslationFunctions { return this.i18nService.LL }
}
```

Passing in the `i18nService` to the constructor and creating a getter feels unnecessary.. is there an option in Angular to have a global object that can be referenced anywhere inside the application?

**Comment by ivanhofer** [comment]
@pit999 did you integrate typesafe-i18n into your Angular application? Did everything work?

**Comment by pit999** [comment]
Hi Ivan, 
sorry for late reply, but I was off some days and this is not my main project (unfortunately :-))

Following code seems to be sufficient too:
import { I18nService } from 'src/i18n/i18n.service';
...
constructor(
    public i18nService: I18nService
  ) {

In Component/ts file:
this.i18nService.LL.deleteMessage();

And in html Code:
{{i18nService.LL.saveButton()}}

**Comment by ivanhofer** [comment]
An official `angular` adapter is now available in version `2.40.0`
